File: Front/src/app/Services/Reservation/reservation.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { Bloc } from 'src/app/models/bloc';
import { Pays } from 'src/app/models/Pays';
import { Residence } from 'src/app/models/Residence';
import { AppartementService } from '../appartement/appartement.service';
import { Appartement } from 'src/app/models/appartement';
import { Reservation } from 'src/app/models/Reservation';

@Injectable({
  providedIn: 'root'
})
export class ReservationService {

  private apiUrl = 'http://localhost:8089/reservation';  

  constructor(private http: HttpClient) { }

  getPays(): Observable<Pays[]> {
    return this.http.get<Pays[]>(`${this.apiUrl}/pays`);
  }

  getResidences(paysId: number): Observable<Residence[]> {
    return this.http.get<Residence[]>(`${this.apiUrl}/residences/${paysId}`);
  }

  getBlocs(residenceId: number): Observable<Bloc[]> {
    return this.http.get<Bloc[]>(`${this.apiUrl}/blocs/${residenceId}`);
  }

  getAppartements(blocId: number): Observable<Appartement[]> {
    return this.http.get<Appartement[]>(`${this.apiUrl}/appartements/${blocId}`);
  }

  createReservation(appartementId: number, start: string, end: string): Observable<Reservation> {
    return this.http.post<Reservation>(`${this.apiUrl}/${appartementId}?start=${start}&end=${end}`, {});
  }

reserverAppartement(appartementId: number, userId: number): Observable<any> {
  return this.http.post(`${this.apiUrl}/reserver/${appartementId}?userId=${userId}`, {}, { responseType: 'text' });
}



checkReservation(appartementId: number, userId: number): Observable<boolean> {
  return this.http.get<boolean>(`${this.apiUrl}/exists?appartementId=${appartementId}&userId=${userId}`);
}


reserverAppartementWithDates(appartementId: number, userId: number, dateDebut: string, dateFin: string) {
  return this.http.post<any>(`${this.apiUrl}/reserver/${appartementId}`, null, {
    params: {
      userId: userId.toString(),
      dateDebut: dateDebut,
      dateFin: dateFin
    }
  });
}

}


 
 

File: Back/src/main/java/esprit/example/fegaac1/services/ReservationService.java
package esprit.example.fegaac1.services;

import esprit.example.fegaac1.entities.*;
import esprit.example.fegaac1.repository.*;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;

import java.time.LocalDate;
import java.util.Collections;
import java.util.List;
import java.util.Set;

@Service
@RequiredArgsConstructor
public class ReservationService  implements IReservationService {



    private final PaysRepository paysRepository;
    private final ResidenceRepository residenceRepository;
    private final BlocRepository blocRepository;
    private final AppartementRepository appartementRepository;
    private final ReservationRepository reservationRepository;
    private final UserRepository userRepository;


    // Get all countries
    public List<Pays> getAllPays() {
        return paysRepository.findAll();
    }

    // Get residences of a specific country
    public Set<Residence> getResidencesByPays(Long paysId) {
        return paysRepository.findById(paysId)
                .map(Pays::getResidences)
                .orElse(Collections.emptySet());
    }

    // Get blocs of a specific residence
    public Set<Bloc> getBlocsByResidence(Long residenceId) {
        return residenceRepository.findById(residenceId)
                .map(Residence::getBlocs)
                .orElse(Collections.emptySet());
    }

    // Get appartements of a specific bloc
    public Set<Appartement> getAppartementsByBloc(Long blocId) {
        return blocRepository.findById(blocId)
                .map(Bloc::getAppartement)
                .orElse(Collections.emptySet());
    }




    public Reservation createReservation(Long appartementId, LocalDate start, LocalDate end) {
        Appartement appartement = appartementRepository.findById(appartementId)
                .orElseThrow(() -> new RuntimeException("Appartement not found"));
        Reservation reservation = new Reservation();
        reservation.setAppartements(appartement);
        reservation.setDateDebut(start);
        reservation.setDateFin(end);
        reservation.setApproved(2); // default
        return reservationRepository.save(reservation);
    }


    public Reservation reserverAppartement(Long appartementId, Long userId, LocalDate dateDebut, LocalDate dateFin) {
        Appartement appartement = appartementRepository.findById(appartementId)
                .orElseThrow(() -> new IllegalArgumentException("Appartement not found with id: " + appartementId));

        User user = userRepository.findById(userId)
                .orElseThrow(() -> new IllegalArgumentException("User not found with id: " + userId));

        // Validate dates
        if (dateDebut == null || dateFin == null || dateDebut.isAfter(dateFin)) {
            throw new IllegalArgumentException("Invalid reservation dates");
        }

        // Check for overlapping reservations
        boolean hasOverlap = hasOverlap(userId, dateDebut, dateFin);
        if (hasOverlap) {
            throw new IllegalArgumentException("You already have a reservation that overlaps with these dates");
        }

        Reservation reservation = new Reservation();
        reservation.setAppartements(appartement);
        reservation.setUser(user);
        reservation.setDateDebut(dateDebut);
        reservation.setDateFin(dateFin);
        reservation.setApproved(2);

        return reservationRepository.save(reservation);
    }




    public boolean existsByUserAndAppartement(Long userId, Long appartementId) {
        return reservationRepository.existsReservation(userId, appartementId);
    }



    public List<Reservation> getReservationsByUser(Long userId) {
        User user = userRepository.findById(userId)
                .orElseThrow(() -> new RuntimeException("User not found"));
        return reservationRepository.findByUser(user);
    }




    public boolean hasApprovedReservation(Long userId) {
        return reservationRepository.existsByUserIdAndApproved(userId, 1);
    }


    public boolean hasOverlap(Long userId, LocalDate newStart, LocalDate newEnd) {
        System.out.println("Backend: checking overlap for user " + userId +
                " from " + newStart + " to " + newEnd);

        List<Reservation> approvedReservations = reservationRepository.findApprovedReservationsByUser(userId);

        for (Reservation r : approvedReservations) {
            System.out.println("Backend: existing approved reservation = " +
                    r.getDateDebut() + " -> " + r.getDateFin());
            if (!(newEnd.isBefore(r.getDateDebut()) || newStart.isAfter(r.getDateFin()))) {
                System.out.println("Backend: overlap detected!");
                return true;
            }
        }
        System.out.println("Backend: no overlap");
        return false;
    }




        public Reservation updateApproval(Long reservationId, int status) {
        Reservation r = reservationRepository.findById(reservationId)
                .orElseThrow(() -> new RuntimeException("Reservation not found"));
        r.setApproved(status);
        return reservationRepository.save(r);
    }








}
